accounts/abi/bind/backends/simulated.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+// SimulatedBackend implements bind.ContractBackend, simulating a blockchain in
+// the background. Its main purpose is to allow easily testing contract bindings.
+type SimulatedBackend struct {
+	database   ethdb.Database   // In memory database to store our testing data
+	blockchain *core.BlockChain // Ethereum blockchain to handle the consensus
+
+	pendingBlock *types.Block   // Currently pending block that will be imported on request
+	pendingState *state.StateDB // Currently pending state that will be the active on on request
+}
+
+// NewSimulatedBackend creates a new binding backend using a simulated blockchain
+// for testing purposes.
+func NewSimulatedBackend(accounts ...core.GenesisAccount) *SimulatedBackend {
+	database, _ := ethdb.NewMemDatabase()
+	core.WriteGenesisBlockForTesting(database, accounts...)
+	blockchain, _ := core.NewBlockChain(database, new(core.FakePow), new(event.TypeMux))
+
+	backend := &SimulatedBackend{
+		database:   database,
+		blockchain: blockchain,
+	}
+	backend.Rollback()
+
+	return backend
+}
+
+// Commit imports all the pending transactions as a single block and starts a
+// fresh new state.
+func (b *SimulatedBackend) Commit() {
+	if _, err := b.blockchain.InsertChain([]*types.Block{b.pendingBlock}); err != nil {
+		panic(err) // This cannot happen unless the simulator is wrong, fail in that case
+	}
+	b.Rollback()
+}
+
+// Rollback aborts all pending transactions, reverting to the last committed state.
+func (b *SimulatedBackend) Rollback() {
+	blocks, _ := core.GenerateChain(b.blockchain.CurrentBlock(), b.database, 1, func(int, *core.BlockGen) {})
+
+	b.pendingBlock = blocks[0]
+	b.pendingState, _ = state.New(b.pendingBlock.Root(), b.database)
+}
+
+// ContractCall implements ContractCaller.ContractCall, executing the specified
+// contract with the given input data.
+func (b *SimulatedBackend) ContractCall(contract common.Address, data []byte, pending bool) ([]byte, error) {
+	// Create a copy of the current state db to screw around with
+	var (
+		block   *types.Block
+		statedb *state.StateDB
+	)
+	if pending {
+		block, statedb = b.pendingBlock, b.pendingState
+	} else {
+		block = b.blockchain.CurrentBlock()
+		statedb, _ = b.blockchain.State()
+	}
+	statedb = statedb.Copy()
+
+	// Set infinite balance to the a fake caller account
+	from := statedb.GetOrNewStateObject(common.Address{})
+	from.SetBalance(common.MaxBig)
+
+	// Assemble the call invocation to measure the gas usage
+	msg := callmsg{
+		from:     from,
+		to:       &contract,
+		gasPrice: new(big.Int),
+		gasLimit: common.MaxBig,
+		value:    new(big.Int),
+		data:     data,
+	}
+	// Execute the call and return
+	vmenv := core.NewEnv(statedb, b.blockchain, msg, block.Header())
```

Comment:
Just a FYI, this function has changed in https://github.com/ethereum/go-ethereum/pull/2141. See: https://github.com/ethereum/go-ethereum/pull/2141/files#diff-0a440a2dbb68b1d118735c02285f90b0R56